packages/jupyter-ai-magics/jupyter_ai_magics/__init__.py

```diff
@@ -1,34 +1,72 @@
+from langchain_core._import_utils import import_attr
+
 from ._version import __version__
 
-# expose embedding model providers on the package root
-from .embedding_providers import (
-    BaseEmbeddingsProvider,
-    GPT4AllEmbeddingsProvider,
-    HfHubEmbeddingsProvider,
-    QianfanEmbeddingsEndpointProvider,
-)
-from .exception import store_exception
-from .magics import AiMagics
-
-# expose JupyternautPersona on the package root
-# required by `jupyter-ai`.
-from .models.persona import JupyternautPersona, Persona
-
-# expose model providers on the package root
-from .providers import (
-    AI21Provider,
-    BaseProvider,
-    GPT4AllProvider,
-    HfHubProvider,
-    QianfanProvider,
-    TogetherAIProvider,
-)
+_dynamic_imports = {
+    # expose embedding model providers on the package root
+    "embedding_providers": [
+        "BaseEmbeddingsProvider",
+        "GPT4AllEmbeddingsProvider",
+        "HfHubEmbeddingsProvider",
+        "QianfanEmbeddingsEndpointProvider",
+    ],
+    "exception": ["store_exception"],
+    "magics": ["AiMagics"],
+    # expose JupyternautPersona on the package root
+    # required by `jupyter-ai`.
+    "models.persona": ["JupyternautPersona", "Persona"],
+    # expose model providers on the package root
+    "providers": [
+        "AI21Provider",
+        "BaseProvider",
+        "GPT4AllProvider",
+        "HfHubProvider",
+        "QianfanProvider",
+        "TogetherAIProvider",
+    ],
+}
```

Comment:
When consumers import from `jupyter_ai_magics`, do they still get type inference for `mypy` & IDE intellisense? I would test this myself but I'm short on time this week.

Reply:
It depends on the editor. The explicit `__all__` helps, but yes not all editros will autocomplete. I think adding typing stubs could help but I am not sure if it is worth the effort.

I have not checked mypy yet.

We could recommend importing from specific modules instead of the top level which will in 100% preserve completions and mypy checks (and in case of base provider be faster too).